remote/remote_state_s3.go

```diff
@@ -341,32 +343,20 @@ func validateS3Config(extendedConfig *ExtendedRemoteStateConfigS3, terragruntOpt
 
 // If the bucket specified in the given config doesn't already exist, prompt the user to create it, and if the user
 // confirms, create the bucket and enable versioning for it.
-func createS3BucketIfNecessary(s3Client *s3.S3, config *ExtendedRemoteStateConfigS3, terragruntOptions *options.TerragruntOptions) error {
-	if !DoesS3BucketExist(s3Client, &config.remoteStateConfigS3) {
-		prompt := fmt.Sprintf("Remote state S3 bucket %s does not exist or you don't have permissions to access it. Would you like Terragrunt to create it?", config.remoteStateConfigS3.Bucket)
-		shouldCreateBucket, err := shell.PromptUserForYesNo(prompt, terragruntOptions)
-		if err != nil {
-			return err
-		}
-
-		if shouldCreateBucket {
-			// Creating the S3 bucket occasionally fails with eventual consistency errors: e.g., the S3 HeadBucket
-			// operation says the bucket exists, but a subsequent call to enable versioning on that bucket fails with
-			// the error "NoSuchBucket: The specified bucket does not exist." Therefore, when creating and configuring
-			// the S3 bucket, we do so in a retry loop with a sleep between retries that will hopefully work around the
-			// eventual consistency issues. Each S3 operation should be idempotent, so redoing steps that have already
-			// been performed should be a no-op.
-			description := fmt.Sprintf("Create S3 bucket %s", config.remoteStateConfigS3.Bucket)
-			maxRetries := 3
-			sleepBetweenRetries := 10 * time.Second
-
-			return util.DoWithRetry(description, maxRetries, sleepBetweenRetries, terragruntOptions.Logger, func() error {
-				return CreateS3BucketWithVersioningSSEncryptionAndAccessLogging(s3Client, config, terragruntOptions)
-			})
-		}
-	}
-
-	return nil
+func createS3BucketWithRetry(s3Client *s3.S3, config *ExtendedRemoteStateConfigS3, terragruntOptions *options.TerragruntOptions) error {
```

Comment:
This change of moving `DoesS3BucketExist` check to `CreateS3Bucket` makes me nervous because it is a fairly significant change in behavior.

Before, terragrunt never attempted to "migrate" an existing bucket to match the configurations and defaults that you have set. This means that even if we update bucket parameters across releases, and if the user upgrades terragrunt versions, the s3 bucket remained untouched. This is somewhat surprising and confusing behavior for folks who are used to terraform, but has a few advantages for us. Off the top of my head:

- It is simpler to maintain, as we didn't really have to worry about backwards compatibility in the s3 bucket creation code: we always started from scratch with a new bucket.

- It doesn't have the behavior of modifying the state bucket during a `plan` or `validate` stage, which can be surprising to a user attempting to do a "read-only" task (NOTE: creating the bucket also happens during these stages, which is arguably just as surprising, but in the create case there is an interactive prompt to confirm the creation that softens the impact).

There is potential that this change can break a lot of existing users, and we also should think through the implications of applying some of the updates that this will do. From the looks of it, this change appears to be orthogonal to the access logging change, so I think this specific update should be in a separate PR where we can discuss the implications of this update.

Reply:
Thank @yorinasub17 This is really useful insight. Here's some of my thoughts, however, based on the concerns you've got, it might be easier to chat through this on a call so I can understand better _(I suspect I'm missing some context & understanding)_

- the `CreateS3Bucket` func has only been called from 2 places, one of which is the new introduction for creating logs bucket, and the other one being `CreateS3BucketWithVersioningSSEncryptionAndAccessLogging`
- the `CreateS3BucketWithVersioningSSEncryptionAndAccessLogging` func is also only called from 1 more place - the `createS3BucketWithRetry` method and following the chain of functioncalls, only when `func (remoteState *RemoteState) Initialize(...)` is called.
- based on the above 2 observations, I'm not following exactly where the concern that this behaviour will impact existing users will come from. As far as I see, this behaviour of `DoesS3BucketExist` has been pushed closer to the actual s3 client API call to create the bucket, but it doesn't ultimately change the behaviour we execute when calling into `terragrunt apply | plan | validate`
- we do get the prompt every time to create the bucket, even if `terragrunt validate` has been called - the user should still be prompted to see if they want the S3 bucket created or not

Reply:
I agree that this change can be extracted into its own separate PR, so I'll raise a separate one for it, so we can have an appropriate discussion.

Reply:
The concern I have is that `CreateS3BucketWithVersioningSSEncryptionAndAccessLogging` will do additional mutating actions to the bucket after `CreateS3Bucket` is called, specifically:

- `EnableEnforcedTLSAccesstoS3Bucket`
- `EnablePublicAccessBlockingForS3Bucket`
- `TagS3Bucket`
- `EnableVersioningForS3Bucket`
- `EnableSSEForS3BucketWide`
- `EnableAccessLoggingForS3BucketWide`

Before, this whole routine was skipped when the bucket already existed, meaning that we did not attempt to mutate that bucket to enable these features. Now, we are only skipping the create routine, meaning that we will attempt to do these mutations on that existing bucket, which may have been created before all of this code existed. My main concern here is that we may add further mutations in this routine (like we are doing now in this PR), and we will need to make sure that those mutations work even if the bucket was created in older versions of terragrunt, potentially as far back as `0.18`.

Reply:
ahaa! got it - yeah, totally missed this! Ok, just about to raise the separate PR for this.

Reply:
> This change of moving `DoesS3BucketExist` check to `CreateS3Bucket` makes me nervous because it is a fairly significant change in behavior.
> 
> Before, terragrunt never attempted to "migrate" an existing bucket to match the configurations and defaults that you have set. This means that even if we update bucket parameters across releases, and if the user upgrades terragrunt versions, the s3 bucket remained untouched. This is somewhat surprising and confusing behavior for folks who are used to terraform, but has a few advantages for us. Off the top of my head:
> 
> * It is simpler to maintain, as we didn't really have to worry about backwards compatibility in the s3 bucket creation code: we always started from scratch with a new bucket.
> * It doesn't have the behavior of modifying the state bucket during a `plan` or `validate` stage, which can be surprising to a user attempting to do a "read-only" task (NOTE: creating the bucket also happens during these stages, which is arguably just as surprising, but in the create case there is an interactive prompt to confirm the creation that softens the impact).
> 
> There is potential that this change can break a lot of existing users, and we also should think through the implications of applying some of the updates that this will do. From the looks of it, this change appears to be orthogonal to the access logging change, so I think this specific update should be in a separate PR where we can discuss the implications of this update.



> here is to disassociate merging this in

So, I just pushed the last bit of changes i've added to this repo, but so far the test is not proving anything, as it points to  the not-implemented function in Terratest yet. 

Do you suggest that I implement a temp function using the s3client here, or is leaving it as a commented `TODO` like you've pointed out enough?